src/metabase/api/session.clj

```diff
@@ -360,7 +360,7 @@
   (api/let-404 [pulse-channel (t2/select-one PulseChannel :pulse_id pulse-id :channel_type "email")]
     (let [emails (get-in pulse-channel [:details :emails])]
       (if (some #{email} emails)
-        (t2/update! PulseChannel (:id pulse-channel) (assoc-in pulse-channel [:details :emails] (remove #{email} emails)))
+        (t2/update! PulseChannel (:id pulse-channel) (update-in pulse-channel [:details :emails] #(remove #{email} %)))
```

Comment:
I don't think this will change anything, since `emails` are the thing you got from `pulse-channel` just before this line of code and there is no updates going on. So the logic stays the same...

Reply:
Ah yes I see I misread how the `update!` is being used, I thought it was acting like a `swap!`. Can you point me at the right way to make this atomic with toucan2? I'll poke around, still new to the library.

Reply:
Something like `(t2/update! PulseChannel id :details ...some way to pass sql to the db...)`, I think. But there is a problem - `details` is a json field (well, it's `text` in my postgres), so the way to remove an email from json could be different between pgsql and mysql?..

Reply:
I was thinking that toucan would have something like `(t2/update! ... swap-fn)` that would just run a select and update within transaction. 

I think just manually wrapping both of these functions in `(t2-with-transaction ...)` is the idiomatic way to fix this?

Reply:
Well, translating Clojure code into an action inside a database sounds like an ambitious goal, so I doubt toucan will ever go there. :)

Yeah, I guess transaction should fix the problem.

Reply:
Looking around at some other APIs it looks like this is indeed the idiom.